### PR TITLE
Bug: gracefully handle bad search params #153833417

### PR DESCRIPTION
**Pull request by weedySeaDragon**
### PT Story:  https://www.pivotaltracker.com/story/show/153833417

This is a WIP because I am not able to come up with "bad" parameters for `ShfApplicationsController #index ` or `UsersController #index`

I can only come up with "bad" parameters for `CompaniesController #index` because we are searching for `ids` .

Can someone else come up with "bad" parameters so that we can really test the `rescue` block for ShfApplicationsController and UsersController?

_These 2 spec tests will *fail:*_
  - ShfApplicationsController bad search parameters....
  - UsersController bad search parameters...

### Changes proposed in this pull request:
1.  (tbd)
2.
3.

Ready for review:
@AgileVentures/shf-project-team

### CODE REVIEW

**Comment by RobertCram** [comment]
@weedySeaDragon: This is how I see the problem: Facebook corrects Rails array params by putting indices in the brackets (preventing duplicate param keys: not entirely unlogical). But this means that Rails will interpret the resulting params as Hashes instead of Arrays, and Ransack chokes on that:

Parameters as they should be:
`Parameters: {"utf8"=>"✓", "q"=>{"business_categories_id_in"=>["2", "2", "2"], "addresses_region_id_in"=>["", "6"], "addresses_kommun_id_in"=>[""], "name_in"=>[""]}, "commit"=>"Sök"}`

Parameters after "correction" by FB:
`Parameters: {"utf8"=>"✓", "q"=>{"business_categories_id_in"=>{"0"=>"2", "1"=>"2", "2"=>"2"}, "addresses_region_id_in"=>{"0"=>"6"}, "addresses_kommun_id_in"=>{"0"=>""}, "name_in"=>{"0"=>""}}, "commit"=>"Sök"}`


For me, the easy fix would be to change the hashes back to arrays before calling Ransack:
```
def fix_facebook_bug(action_params){
  action_params.each_pair {| key, value | q[key] = value.values if value.class == Hash}
end
```

I've tried it, and it works like a charm.

**Comment by RobertCram** [comment]
@weedySeaDragon: This is how I see the problem: Facebook corrects Rails array params by putting indices in the brackets (preventing duplicate param keys: not entirely unlogical). But this means that Rails will interpret the resulting params as Hashes instead of Arrays, and Ransack chokes on that:

Parameters as they should be:
`Parameters: {"utf8"=>"✓", "q"=>{"business_categories_id_in"=>["2", "2", "2"], "addresses_region_id_in"=>["", "6"], "addresses_kommun_id_in"=>[""], "name_in"=>[""]}, "commit"=>"Sök"}`

Parameters after "correction" by FB:
`Parameters: {"utf8"=>"✓", "q"=>{"business_categories_id_in"=>{"0"=>"2", "1"=>"2", "2"=>"2"}, "addresses_region_id_in"=>{"0"=>"6"}, "addresses_kommun_id_in"=>{"0"=>""}, "name_in"=>{"0"=>""}}, "commit"=>"Sök"}`


For me, the easy fix would be to change the hashes back to arrays before calling Ransack:
```
def fix_facebook_bug(action_params){
  action_params.each_pair {| key, value | q[key] = value.values if q.to_unsafe_h[key].is_a?(Hash)}
end
```

Specs & features pass and couldn't find any other negative effects (but YMMV of course)

**Comment by RobertCram** [comment]
@weedySeaDragon As to your question to come up with "bad" parameters for the other controllers: that's gonna be a problem, because FB can do it's correction without it leading to an exception, while still giving the wrong results. E.g. the next URL (shf_application, searching for two company numbers):

p.s. You have to be logged in for the links below to work

Correct Link:
https://hitta.sverigeshundforetagare.se/en/ansokan?utf8=%E2%9C%93&q[user_membership_number_in][]&q[user_last_name_in][]&q[company_number_in][]&q[company_number_in][]=4704211681&q[company_number_in][]=4806085686&q[state_in][]&commit=Search


Facebook corrected link:
https://hitta.sverigeshundforetagare.se/en/ansokan?utf8=%E2%9C%93&q%5Buser_membership_number_in%5D%5B0%5D&q%5Buser_last_name_in%5D%5B0%5D&q%5Bcompany_number_in%5D%5B0%5D&q%5Bcompany_number_in%5D%5B1%5D=4704211681&q%5Bcompany_number_in%5D%5B2%5D=4806085686&q%5Bstate_in%5D%5B0%5D&commit=Search

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -77,4 +83,29 @@ def prepare_exception_notifier
   end
 
 
+  # When one of our URLS is cut and pasted in Facebook (FB), FB will actually
+  # _change_ the URL.  Then anyone in FB clicking on the URL will be taken
+  # to our system with a (possibly) invalid URL.
+  # Ex:  I want to share the list of dog groomers in Skäne (a region) with someone.
+  #   I do a search for all companies in Skäne with dog groomers.  I copy that resulting URL
```
Comment (thesuss) [inline app/controllers/application_controller.rb]: Skåne, but I guess that is totally not needed to change =)

**Review by thesuss — APPROVED** [review]

File: spec/controllers/companies_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    it 'empty values do not need to be retained)' do
+
+      fb_mangled_params = { "utf8" => "✓",
+                            "q" => {
+                                "business_categories_id_in" => { "0" => "2", "1" => "2", "2" => "2" },
+                                "addresses_region_id_in" => { "0" => "6" },
+                                "addresses_kommun_id_in" => { "0" => "" },
+                                "name_in" => { "0" => "" } },
+                            "commit" => "Sök" }
```
Comment (patmbolger) [inline spec/controllers/companies_controller_spec.rb]: I think the pattern we've seen is like this .... 

What we would see without mangling:
```
Parameters: {"utf8"=>"✓", "q"=>{"business_categories_id_in"=>["", "2"], "addresses_region_id_in"=>["", "6"], "addresses_kommun_id_in"=>[""], "name_in"=>[""]}, "commit"=>"Sök"}
```
versus the mangled version:
```
Parameters: {"utf8"=>"✓", "q"=>{"business_categories_id_in"=>{"0"=>nil, "1"=>"2"}, "addresses_region_id_in"=>{"0"=>nil, "1"=>"6"}, "addresses_kommun_id_in"=>{"0"=>nil}, "name_in"=>{"0"=>nil}}, "commit"=>"Sök"}
```
That is, we don't see the occurrence of `{ "0" => ""  }` but rather `{ "0"=>nil }`.

So perhaps a more realistic version of a mangled test hash is:
```
{ "utf8" => "✓",
                            "q" => {
                                "business_categories_id_in" => { "0" => "1", "1" => "2", "2" => "3" },
                                "addresses_region_id_in" => { "0" => "6" },
                                "addresses_kommun_id_in" => { "0" => nil },
                                "name_in" => { "0" => nil } },
                            "commit" => "Sök" }
```
(note that I also changed the cat ID's to be three different values, since we'd never really see the same value 3 times).

File: spec/controllers/controller_shared_examples.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+RSpec.shared_examples 'it gracefully handles bad search parameters' do |test_description, bad_parameters, expected_redirect, plural_search_items|
```
Comment (patmbolger) [inline spec/controllers/controller_shared_examples.rb]: Did you mean to use this in the tests below?  Or, perhaps it is intended to be used for another PR in which we rescue the error produced by bad params and redirect to some place?

**Review by patmbolger — COMMENTED** [review]
Looks good.  Please see comments.

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -77,4 +83,29 @@ def prepare_exception_notifier
...
+  #   and share it with my friends in FB.  One of my friends clicks on it but
+  #   because FB has mangled the URL, it no longer works and causes a 500 or 404 error.
+  #
+  # Specifically, FB may change  a value that should be an Array to a Hash.  this changes it back.
+  # This is important for the Ransack gem, which we use for searching and sorting in some views.
+  # (Search for the `.ransack` method in controllers.)
+  # Ransack expects the parameters with the 'q' key to have values that are Arrays, not Hashes.
```
Comment (patmbolger) [inline app/controllers/application_controller.rb]: It's pretty clear that FB is converting our params to JSON, and then just serving that JSON right back to us without backing out the conversion.  This may or may not be worth mentioning here.

File: spec/controllers/controller_shared_examples.rb (hunk 3 above)
Comment (weedySeaDragon) [reply spec/controllers/controller_shared_examples.rb]: it was left over from my first version.  deleted now.

**Review by patmbolger — APPROVED** [review]

**Review by RobertCram — APPROVED** [review]
